book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,274 @@
+}
+
+trait AnchorVariant {
+    type Shared;
+    type PerSpend;
+}
+```
+
+### Changes to Sapling ShieldedData
+[changes-to-sapling-shieldeddata]: #changes-to-sapling-shieldeddata
+
+We use `AnchorVariant` in `ShieldedData` to model the anchor differences between V4 and V5:
+
+```rust
+struct sapling::ShieldedData<AnchorV: AnchorVariant> {
+    value_balance: Amount,
+    anchor: AnchorV::Shared,
+    first: Either<Spend<AnchorV>, Output>,
+    rest_spends: Vec<Spend<AnchorV>>,
+    rest_outputs: Vec<Output>,
+    binding_sig: Signature<Binding>,
+}
+```
+
+### Adding V5 Sapling Spend
+[adding-v5-sapling-spend]: #adding-v5-sapling-spend
+
+Sapling spend code is located at `zebra-chain/src/sapling/spend.rs`. We use `AnchorVariant` to model the anchor differences between V4 and V5:
+
+```rust
+struct Spend<AnchorV: AnchorVariant> {
+    cv: commitment::ValueCommitment,
+    anchor: AnchorV::PerSpend,
+    nullifier: note::Nullifier,
+    rk: redjubjub::VerificationKeyBytes<SpendAuth>,
+    zkproof: Groth16Proof,
+    spend_auth_sig: redjubjub::Signature<SpendAuth>,
+}
+```
+
+### No Changes to Sapling Output
+[no-changes-to-sapling-output]: #no-changes-to-sapling-output
+
+In Zcash the Sapling output representations are the same for V4 and V5 transactions, so no variants are needed. The output code is located at `zebra-chain/src/sapling/output.rs`:
+
+```rust
+struct Output {
+    cv: commitment::ValueCommitment,
+    cm_u: jubjub::Fq,
+    ephemeral_key: keys::EphemeralPublicKey,
+    enc_ciphertext: note::EncryptedNote,
+    out_ciphertext: note::WrappedNoteKey,
+    zkproof: Groth16Proof,
+}
+```
+
+## Orchard Additions
+[orchard-additions]: #orchard-additions
+
+### Adding V5 Transactions
+[adding-v5-transactions]: #adding-v5-transactions
+
+Now lets see how the V5 transaction is specified in the protocol, this is the second table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus) and how are we going to represent it based in the above changes for Sapling fields and the new Orchard fields.
+
+We propose the following representation for transaction V5 in Zebra:
+
+```rust
+enum Transaction::V5 {
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    sapling_shielded_data: Option<sapling::ShieldedData<SharedAnchor>>,
+    orchard_shielded_data: Option<orchard::ShieldedData>,
+}
+```
+
+To model the V5 anchor type, `sapling_shielded_data` uses the `SharedAnchor` variant located at `zebra-chain/src/transaction/sapling/shielded_data.rs`.
+
+### Adding Orchard ShieldedData
+[adding-orchard-shieldeddata]: #adding-orchard-shieldeddata
+
+The new V5 structure will create a new `orchard::ShieldedData` type. This new type will be defined in a new `zebra-chain/src/orchard/shielded_data.rs` file:
+
+```rust
+struct orchard::ShieldedData {
+    flags: Flags,
+    value_balance: Amount,
+    anchor: tree::Root,
```

Comment:
Non-blocking: should this be an `SharedAnchor`? For explicitness/consistency?

Reply:
Here are some design alternatives:
```suggestion
    shared_anchor: tree::Root,
```

The syntax here is much harder to read and understand:
```suggestion
    shared_anchor: <SharedAnchor as AnchorVariant>::Shared,
```